# Proyek Analisis Data: [Brazilian E-Commerce Public Dataset by Olist]
- **Nama:** Yunico Ardian Pradana
- **Email:** yunicoardian123@gmail.com
- **ID Dicoding: yunicoardian** 

## Menentukan Pertanyaan Bisnis
- Pertanyaan 1 : Negara bagian mana saja yang memiliki pendapatan tertinggi?
- Pertanyaan 2 : Bagaimana Trend Pendapatan/Pemasukan di 5 negara dengan revenue tertinggi ?
- Pertanyaan 3 : Produk apa saja yang menjadi sumber pendapatan terbesar di lima negara bagian tersebut ?

## Import Semua Packages/Library yang Digunakan

In [1]:
import pandas as pd
import zipfile
import plotly.express as px
from plotly.subplots import make_subplots 

## Data Wrangling

### Gathering Data

In [2]:
# read zip
zip_file = zipfile.ZipFile('archive.zip', 'r')
# extract zip to new folder
zip_file.extractall('datasets/')
# close zipfile
zip_file.close()

#### Baca Dataframe customers

In [3]:
# df_customer = pd.read_csv('datasets/olist_customers_dataset.csv')
df_customer = pd.read_csv('datasets/olist_customers_dataset.csv')
df_customer.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

#### Baca Dataframe geolocation

In [4]:
df_geolocation = pd.read_csv('datasets/olist_geolocation_dataset.csv')
df_geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

#### Baca Dataframe order_items

In [5]:
df_order_items = pd.read_csv('datasets/olist_order_items_dataset.csv')
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

#### Baca Dataframe order_payments

In [6]:
df_order_payments = pd.read_csv('datasets/olist_order_payments_dataset.csv')
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

#### Baca Dataframe order_reviews

In [7]:
df_order_reviews = pd.read_csv('datasets/olist_order_reviews_dataset.csv')
df_order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

#### Baca Dataframe orders_dataset

In [8]:
df_orders_dataset = pd.read_csv('datasets/olist_orders_dataset.csv')
df_orders_dataset.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

#### Baca Dataframe products_dataset

In [9]:
df_products_dataset = pd.read_csv('datasets/olist_products_dataset.csv')
df_products_dataset.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

#### Baca Dataframe sellers_dataset

In [10]:
df_sellers_dataset = pd.read_csv('datasets/olist_sellers_dataset.csv')
df_sellers_dataset.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

#### Baca product category

In [11]:
df_product_name = pd.read_csv('datasets/product_category_name_translation.csv')
df_product_name.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

### Assessing Data

In [12]:
# Store dataframes in a list
dataframes = [df_customer, df_geolocation, df_order_items, df_order_payments, df_order_reviews, df_orders_dataset, df_products_dataset, df_sellers_dataset, df_product_name]

for df in dataframes:
    # Get the variable name using globals()
    df_name = [name for name, obj in globals().items() if obj is df][0]
    print(f"\nCustom Info for {df_name} dataframe:")
    print(df.info())


Custom Info for df_customer dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None

Custom Info for df_geolocation dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng    

Selanjutnya kita mencoba untuk melihat jumlah Null Value pada tiap dataframe

In [13]:
for df in dataframes:
    # Get the variable name using globals()
    df_name = [name for name, obj in globals().items() if obj is df][0]
    
    print(f"\nMissing Values in {df_name} dataframe:")
    print(df.isna().sum())


Missing Values in df_customer dataframe:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Missing Values in df_geolocation dataframe:
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

Missing Values in df_order_items dataframe:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Missing Values in df_order_payments dataframe:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Missing Values in df_order_reviews dataframe:
review_id                      0
order_id                       0
review_score                   0
re

selanjutnya kita cek juga duplikat pada tiap dataframe

In [14]:
for df in dataframes:
    # Get the variable name using globals()
    df_name = [name for name, obj in globals().items() if obj is df][0]
    print(f"\nTotal baris duplikat pada {df_name} dataframe: {df.duplicated().sum()}")


Total baris duplikat pada df_customer dataframe: 0

Total baris duplikat pada df_geolocation dataframe: 261831

Total baris duplikat pada df_order_items dataframe: 0

Total baris duplikat pada df_order_payments dataframe: 0

Total baris duplikat pada df_order_reviews dataframe: 0

Total baris duplikat pada df_orders_dataset dataframe: 0

Total baris duplikat pada df_products_dataset dataframe: 0

Total baris duplikat pada df_sellers_dataset dataframe: 0

Total baris duplikat pada df_product_name dataframe: 0


selanjutnya check distribusi data

In [15]:
for df in dataframes:
    # Get the variable name using globals()
    df_name = [name for name, obj in globals().items() if obj is df][0]
    
    print(f"\nDistribusi data pada {df_name} :")
    print(df.describe())


Distribusi data pada df_customer :
       customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

Distribusi data pada df_geolocation :
       geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02

Distribusi data

In [16]:
df_orders_dataset.order_status.value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Cleaning Data

Pada saat kita check duplikat terlihat bahwa di dataframe df_geolocation terdapat duplikat sejumlah 261831 sehingga perlu penanganan lebih lanjut dengan cara menghapus diplikat

In [17]:
df_geolocation.drop_duplicates(inplace=True)

setelah kita menghapus duplikat, cek kembali untuk memastikan tidak ada duplikat 

In [18]:
df_geolocation.duplicated().sum()

0

selanjutnya ialah kita check null value pada dataframe df_order_reviews

In [19]:
df_order_reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Hasil diatas menunjukkan kolom review_comment_title dan review_comment_message terdapat nilai null value, maka langkah selanjutnya ialah kita melakukan imputasi atau menambahkan nilai default string "No Tittle" dan "No Message"

In [20]:
df_order_reviews["review_comment_message"].fillna("No Message", inplace=True)
df_order_reviews["review_comment_title"].fillna("No Title", inplace=True)

Pada saat kita assessing data, terdapat beberapa kolom yang seharusnya tipe datanya datetime malah obj/str. oleh karena itu kita perlu mengubahnya kedalah format datetime. beberapa kolom yang memerlukan perubahan ada di dataset df_orders_dataset, df_order_reviews, dan df_order_items

In [21]:
df_order_items['shipping_limit_date'] = pd.to_datetime(df_order_items['shipping_limit_date'])

In [22]:
orders_column = [
    'order_delivered_carrier_date', 
    'order_delivered_customer_date', 
    'order_estimated_delivery_date', 
    'order_purchase_timestamp', 
    'order_approved_at'
]

for column in orders_column:
    df_orders_dataset[column] = pd.to_datetime(df_orders_dataset[column])

In [23]:
col123 = ['review_creation_date', 'review_answer_timestamp']
df_order_reviews[col123] = df_order_reviews[col123].apply(pd.to_datetime)

## Exploratory Data Analysis (EDA)

### Pertanyaan No 1

In [24]:
# Hanya pesanan yang telah dikirim yang akan digunakan dalam analisis berikut.
df_orders123 = df_orders_dataset[df_orders_dataset['order_status'] == 'delivered'] # Pilih pesanan dengan status terkirim
df_orders123 = df_orders123.reset_index(drop = True) # Mengatur ulang indeks kolom

# Dalam analisis berikut ini ada 4 kolom yang akan digunakan dari tabel orders: 
# order_id, customer_id, order_status, order_purchase_timestamp

kolom_yang_dihapus = ['order_approved_at','order_delivered_carrier_date', 'order_delivered_customer_date', 
'order_estimated_delivery_date']
df_orders123 = df_orders123.drop(kolom_yang_dihapus, axis=1)

df_orders123['order_purchase_timestamp'] = pd.to_datetime(df_orders123['order_purchase_timestamp'])
df_orders123.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96478 entries, 0 to 96477
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   order_id                  96478 non-null  object        
 1   customer_id               96478 non-null  object        
 2   order_status              96478 non-null  object        
 3   order_purchase_timestamp  96478 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(3)
memory usage: 2.9+ MB


In [25]:
# Gabungkan dataframe df_orders123 dan df, lalu simpan di dataframe revenue_states
revenue_states = pd.merge(df_orders123, df_customer, how='left', on='customer_id')

# Gabungkan dataframe revenue_states dan payments table, then stored in revenue_states
revenue_states = pd.merge(revenue_states, df_order_payments, how='left', on='order_id')
revenue_states

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
...                                  ...                               ...   
100752  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
100753  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
100754  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
100755  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
100756  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp  \
0         delivered      2017-10-02 10:56:33   
1         delivered      2017-10-02 10:56:33   
2         delivered      2017-10-02 10:56:33   
3         delivered      2018-07-24 20:41:37   
4         delivered      2018-08-08 08:38:49   
...             ...                      ...   
100752    delivered      2017-03-09 09:54:05   
100753    delivered      2018-02-06 12:58:58   
100754    delivered      2017-08-27 14:46:43   
100755    delivered      2018-01-08 21:28:27   
100756    delivered      2018-03-08 20:57:30   

                      customer_unique_id  customer_zip_code_prefix  \
0       7c396fd4830fd04220f754e42b4e5bff                      3149   
1       7c396fd4830fd04220f754e42b4e5bff                      3149   
2       7c396fd4830fd04220f754e42b4e5bff                      3149   
3       af07308b275d755c9edb36a90c618231                     47813   
4       3a653a41f6f9fc3d2a113cf8398680e8                     75265   
...                                  ...                       ...   
100752  6359f309b166b0196dbf7ad2ac62bb5a                     12209   
100753  da62f9e57a76d978d02ab5362c509660                     11722   
100754  737520a9aad80b3fbbdad19b66b37b30                     45920   
100755  5097a5312c8b157bb7be58ae360ef43c                     28685   
100756  60350aa974b26ff12caad89e55993bd6                     83750   

              customer_city customer_state  payment_sequential payment_type  \
0                 sao paulo             SP                 1.0  credit_card   
1                 sao paulo             SP                 3.0      voucher   
2                 sao paulo             SP                 2.0      voucher   
3                 barreiras             BA                 1.0       boleto   
4                vianopolis             GO                 1.0  credit_card   
...                     ...            ...                 ...          ...   
100752  sao jose dos campos             SP                 1.0  credit_card   
100753         praia grande             SP                 1.0  credit_card   
100754          nova vicosa             BA                 1.0  credit_card   
100755              japuiba             RJ                 1.0  credit_card   
100756                 lapa             PR                 1.0   debit_card   

        payment_installments  payment_value  
0                        1.0          18.12  
1                        1.0           2.00  
2                        1.0          18.59  
3                        1.0         141.46  
4                        3.0         179.12  
...                      ...            ...  
100752                   3.0          85.08  
100753                   3.0         195.00  
100754                   5.0         271.01  
100755                   4.0         441.16  
100756                   1.0          86.86  

[100757 rows x 12 columns]

ubah singkatan yang ada pada kolom customer_state

In [26]:
# Kamus untuk mengganti singkatan dengan nama lengkap
state_mapping = {
    'AC': 'Acre',
    'AL': 'Alagoas',
    'AM': 'Amazonas',
    'AP': 'Amapa',
    'BA': 'Bahia',
    'CE': 'Ceara',
    'DF': 'Distrito Federal',
    'ES': 'Espirito Santo',
    'GO': 'Goias',
    'MA': 'Maranhao',
    'MG': 'Minas Gerais',
    'MS': 'Mato Grosso do Sul',
    'MT': 'Mato Grosso',
    'PA': 'Para',
    'PB': 'Paraiba',
    'PE': 'Pernambuco',
    'PI': 'Piaui',
    'PR': 'Parana',
    'RJ': 'Rio de Janeiro',
    'RN': 'Rio Grande do Norte',
    'RO': 'Rondonia',
    'RR': 'Roraima',
    'RS': 'Rio Grande do Sul',
    'SC': 'Santa Catarina',
    'SE': 'Sergipe',
    'SP': 'Sao Paulo',
    'TO': 'Tocantins'
}

# Mengganti singkatan dengan nama lengkap menggunakan map
revenue_states['customer_state'] = revenue_states['customer_state'].map(state_mapping)

In [27]:
revenue_states.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100757 entries, 0 to 100756
Data columns (total 12 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   order_id                  100757 non-null  object        
 1   customer_id               100757 non-null  object        
 2   order_status              100757 non-null  object        
 3   order_purchase_timestamp  100757 non-null  datetime64[ns]
 4   customer_unique_id        100757 non-null  object        
 5   customer_zip_code_prefix  100757 non-null  int64         
 6   customer_city             100757 non-null  object        
 7   customer_state            100757 non-null  object        
 8   payment_sequential        100756 non-null  float64       
 9   payment_type              100756 non-null  object        
 10  payment_installments      100756 non-null  float64       
 11  payment_value             100756 non-null  float64       
dtypes:

In [28]:
revenue_states[revenue_states.isna().any(axis=1)]

order_id                       customer_id  \
31175  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   

      order_status order_purchase_timestamp                customer_unique_id  \
31175    delivered      2016-09-15 12:16:38  830d5b7aaa3b6f1e9ad63703bec97d23   

       customer_zip_code_prefix         customer_city customer_state  \
31175                     14600  sao joaquim da barra      Sao Paulo   

       payment_sequential payment_type  payment_installments  payment_value  
31175                 NaN          NaN                   NaN            NaN

Karena hanya ada 4 nilai yang hilang dalam 1 baris dari 100757 baris, menghapus satu baris tidak akan mempengaruhi analisis secara signifikan.

In [29]:
revenue_states = revenue_states.dropna()

In [30]:
# From table 1, group by 'customer_state' and sum the 'payment_value' for each state
revenue_by_state = revenue_states.groupby('customer_state')['payment_value'].sum().reset_index()

# Sort the summary by highest revenue in descending order
revenue_by_state = revenue_by_state.sort_values(by='payment_value', ascending=False)
revenue_by_state.columns = ['State', 'Revenue']
revenue_by_state

State     Revenue
24            Sao Paulo  5770266.19
20       Rio de Janeiro  2055690.45
12         Minas Gerais  1819277.61
19    Rio Grande do Sul   861802.40
15               Parana   781919.55
23       Santa Catarina   595208.40
4                 Bahia   591270.60
6      Distrito Federal   346146.17
8                 Goias   334294.22
7        Espirito Santo   317682.65
16           Pernambuco   309074.59
5                 Ceara   266463.97
13                 Para   212027.55
10          Mato Grosso   181441.72
9              Maranhao   147807.29
14              Paraiba   137834.65
11   Mato Grosso do Sul   134421.54
17                Piaui   105272.17
18  Rio Grande do Norte   100728.30
1               Alagoas    94195.79
25              Sergipe    70289.13
26            Tocantins    60007.37
21             Rondonia    56975.70
3              Amazonas    27596.18
0                  Acre    19586.25
2                 Amapa    16141.81
22              Roraima     9039.52

### Pertanyaan No 2

di pertanyaan ke 2 ini kita ingin melihat trend revenue per bulan dari tahun 2016 hingga 2018 di 5 kota dengan revenue tertinggi. adapun implementasinya ada pada kode dibawah

In [31]:
def calculate_monthly_revenue(revenue_states, state_column, states=None):
    if states:
        revenue_states = revenue_states[revenue_states[state_column].isin(states)]

    revenue_states['year'] = revenue_states['order_purchase_timestamp'].dt.year
    revenue_states['month'] = revenue_states['order_purchase_timestamp'].dt.month
    group_columns = ['year', 'month']
    if state_column in revenue_states.columns:
        group_columns.append(state_column)
    monthly_revenue = revenue_states.groupby(group_columns)['payment_value'].sum().reset_index()
    monthly_revenue = monthly_revenue.sort_values(by=[state_column, 'year', 'month']).reset_index(drop=True)
    monthly_revenue['year_month'] = monthly_revenue['year'].astype(str) + '-' + monthly_revenue['month'].astype(str)
    monthly_revenue.columns = ['year', 'month', 'state', 'revenue', 'year_month']
    return monthly_revenue

# Calculate monthly revenue for specific states
states_to_calculate = ['Sao Paulo', 'Rio de Janeiro', 'Minas Gerais', 'Rio Grande do Sul', 'Parana']
monthly_revenue_states = {}
for state in states_to_calculate:
    monthly_revenue_states[state] = calculate_monthly_revenue(revenue_states, 'customer_state', states=[state])

# Access the results
monthly_revenue_SP = monthly_revenue_states['Sao Paulo']
monthly_revenue_RJ = monthly_revenue_states['Rio de Janeiro']
monthly_revenue_MG = monthly_revenue_states['Minas Gerais']
monthly_revenue_RS = monthly_revenue_states['Rio Grande do Sul']
monthly_revenue_PR = monthly_revenue_states['Parana']


C:\Users\USER\AppData\Local\Temp\ipykernel_11304\3260156733.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  revenue_states['year'] = revenue_states['order_purchase_timestamp'].dt.year
C:\Users\USER\AppData\Local\Temp\ipykernel_11304\3260156733.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  revenue_states['month'] = revenue_states['order_purchase_timestamp'].dt.month
C:\Users\USER\AppData\Local\Temp\ipykernel_11304\3260156733.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a 

In [32]:
monthly_revenue_SP

year  month      state    revenue year_month
0   2016     10  Sao Paulo   13559.93    2016-10
1   2017      1  Sao Paulo   44838.14     2017-1
2   2017      2  Sao Paulo   84444.00     2017-2
3   2017      3  Sao Paulo  147665.99     2017-3
4   2017      4  Sao Paulo  138822.04     2017-4
5   2017      5  Sao Paulo  195065.48     2017-5
6   2017      6  Sao Paulo  191008.60     2017-6
7   2017      7  Sao Paulo  204486.52     2017-7
8   2017      8  Sao Paulo  219219.25     2017-8
9   2017      9  Sao Paulo  238785.05     2017-9
10  2017     10  Sao Paulo  249924.09    2017-10
11  2017     11  Sao Paulo  401027.72    2017-11
12  2017     12  Sao Paulo  313129.20    2017-12
13  2018      1  Sao Paulo  418496.08     2018-1
14  2018      2  Sao Paulo  350994.83     2018-2
15  2018      3  Sao Paulo  429115.21     2018-3
16  2018      4  Sao Paulo  443437.16     2018-4
17  2018      5  Sao Paulo  478505.73     2018-5
18  2018      6  Sao Paulo  392066.82     2018-6
19  2018      7  Sao Paulo  371904.30     2018-7
20  2018      8  Sao Paulo  443770.05     2018-8

### Pertanyaan No 3

pada pertanyaan no 3, kita ingin melihat produk apa saja yang menjadi sumber pendapatan dari 5 negara bagian tersebut. prosesnya melibatkan beberapa penggabungan dataset. adapun implementasinya ada pada kode dibawah ini 

In [33]:
# Merge df_products_dataset table and df_product_name table, stored in products_merged
products_merged = pd.merge(df_products_dataset, df_product_name, how='left', on='product_category_name')

# Merge revenue_states table and df_order_items table, stored in revenue_order_items
revenue_order_items = pd.merge(revenue_states, df_order_items, how='left', on='order_id')

# Merge revenue_order_items table and products_merged table, stored in revenue_order_items
revenue_products = pd.merge(revenue_order_items, products_merged, how='left', on='product_id')

revenue_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115035 entries, 0 to 115034
Data columns (total 27 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       115035 non-null  object        
 1   customer_id                    115035 non-null  object        
 2   order_status                   115035 non-null  object        
 3   order_purchase_timestamp       115035 non-null  datetime64[ns]
 4   customer_unique_id             115035 non-null  object        
 5   customer_zip_code_prefix       115035 non-null  int64         
 6   customer_city                  115035 non-null  object        
 7   customer_state                 115035 non-null  object        
 8   payment_sequential             115035 non-null  float64       
 9   payment_type                   115035 non-null  object        
 10  payment_installments           115035 non-null  float64       
 11  

In [34]:
revenue_products = revenue_products.rename(columns={'product_category_name_english': 'product_category'})

In [35]:
revenue_products = revenue_products[['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'customer_state', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value', 'product_id', 'product_category']]
revenue_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115035 entries, 0 to 115034
Data columns (total 11 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   order_id                  115035 non-null  object        
 1   customer_id               115035 non-null  object        
 2   order_status              115035 non-null  object        
 3   order_purchase_timestamp  115035 non-null  datetime64[ns]
 4   customer_state            115035 non-null  object        
 5   payment_sequential        115035 non-null  float64       
 6   payment_type              115035 non-null  object        
 7   payment_installments      115035 non-null  float64       
 8   payment_value             115035 non-null  float64       
 9   product_id                115035 non-null  object        
 10  product_category          113384 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 9.7+ MB

In [36]:
missing_values = revenue_products.isna().sum().reset_index().rename(columns={'index': 'column_name', 0: 'value'})
missing_values

column_name  value
0                   order_id      0
1                customer_id      0
2               order_status      0
3   order_purchase_timestamp      0
4             customer_state      0
5         payment_sequential      0
6               payment_type      0
7       payment_installments      0
8              payment_value      0
9                 product_id      0
10          product_category   1651

In [37]:
kl = revenue_products[revenue_products.isna().any(axis=1)]
kl

order_id                       customer_id  \
126     bfe42c22ecbf90bc9f35cf591270b6a7  803ac05904124294f8767894d6da532b   
215     58ac1947c1a9067b9f416cba6d844a3f  ee8e1d37f563ecc11cc4dcb4dfd794c2   
268     e22b71f6e4a481445ec4527cb4c405f7  1faf89c8f142db3fca6cf314c51a37b6   
309     a094215e786240fcfefb83d18036a1cd  86acfb656743da0c113d176832c9d535   
381     6f841dde94727854eaff3f66432c80ea  a9c9532060c9d245f06526c633d2dfba   
...                                  ...                               ...   
114603  1a10e938a1c7d8e5eecc3380f71ca76b  8a81607347c25d881d995d94de6ad824   
114773  e33865519137f5737444109ae8438633  64b086bdcc54458af3ea3bd838db54a5   
114780  f0dd9af88d8ef5a8e4670fbbedaf19c4  30ddb50bd22ee927ebe308ea3da60735   
114786  272874573723eec18f23c0471927d778  48e080c8001e92ebea2b64e474f91a60   
114805  dff2b9b8d7cfc595836945e1443789c3  2436fb2666a65fbacae82532e797cabf   

       order_status order_purchase_timestamp  customer_state  \
126       delivered      2018-01-27 22:04:34    Minas Gerais   
215       delivered      2017-09-13 09:18:50       Sao Paulo   
268       delivered      2017-04-22 13:48:18    Minas Gerais   
309       delivered      2018-02-08 18:56:45       Sao Paulo   
381       delivered      2018-01-02 19:20:35  Rio de Janeiro   
...             ...                      ...             ...   
114603    delivered      2018-07-25 08:58:35  Rio de Janeiro   
114773    delivered      2018-05-28 00:44:06       Sao Paulo   
114780    delivered      2017-09-02 20:38:29       Sao Paulo   
114786    delivered      2017-12-20 23:10:33       Sao Paulo   
114805    delivered      2018-07-16 12:59:02  Rio de Janeiro   

        payment_sequential payment_type  payment_installments  payment_value  \
126                    1.0  credit_card                   1.0          42.40   
215                    1.0  credit_card                   5.0         124.16   
268                    1.0  credit_card                   3.0          34.24   
309                    1.0  credit_card                   1.0         132.05   
381                    1.0  credit_card                   3.0         192.95   
...                    ...          ...                   ...            ...   
114603                 1.0       boleto                   1.0         146.19   
114773                 1.0       boleto                   1.0          42.14   
114780                 1.0       boleto                   1.0         100.25   
114786                 1.0  credit_card                   5.0         132.05   
114805                 1.0       boleto                   1.0         235.12   

                              product_id product_category  
126     71225f49be70df4297892f6a5fa62171              NaN  
215     9820e04e332fc450d6fd975befc1bc28              NaN  
268     3bc5164bc7f4be77002d6651da65c98c              NaN  
309     5a848e4ab52fd5445cdc07aab1c40e48              NaN  
381     c600d7f13104e8db2ca2b9fa78581409              NaN  
...                                  ...              ...  
114603  8d39a63db6a98e40571f486e0c711cf7              NaN  
114773  e9cbc0910ab050cbd92fbeb051c270ea              NaN  
114780  0c877471e35049402bca235e06cb8063              NaN  
114786  5a848e4ab52fd5445cdc07aab1c40e48              NaN  
114805  bfa0e398bef2d28bb421d42997982cab              NaN  

[1651 rows x 11 columns]

karena banyak sekali data yang missing value, disini saya memutuskan untuk menghapus baris data dimana category bernilai NaN

In [38]:
revenue_products = revenue_products.dropna()

In [39]:
# menampilkan baris yang duplikat
revenue_products[revenue_products.duplicated(keep=False)]

order_id                       customer_id  \
12      e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
13      e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
57      9faeb9b2746b9d7526aef5acb08e2aa0  79183cd650e2bb0d475b0067d45946ac   
58      9faeb9b2746b9d7526aef5acb08e2aa0  79183cd650e2bb0d475b0067d45946ac   
63      66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   
...                                  ...                               ...   
114985  b3112ca67f3afd4e20cf2ee91fc4f804  6f83c71b6c044fb156d697d4130fe9b5   
115003  e8fd20068b9f7e6ec07068bb7537f781  609b9fb8cad4fe0c7b376f77c8ab76ad   
115004  e8fd20068b9f7e6ec07068bb7537f781  609b9fb8cad4fe0c7b376f77c8ab76ad   
115032  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
115033  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   

       order_status order_purchase_timestamp     customer_state  \
12        delivered      2017-05-16 19:41:10     Rio de Janeiro   
13        delivered      2017-05-16 19:41:10     Rio de Janeiro   
57        delivered      2018-07-26 14:39:59  Rio Grande do Sul   
58        delivered      2018-07-26 14:39:59  Rio Grande do Sul   
63        delivered      2018-03-09 14:50:15            Sergipe   
...             ...                      ...                ...   
114985    delivered      2018-08-02 22:46:54          Sao Paulo   
115003    delivered      2017-08-10 21:21:07          Sao Paulo   
115004    delivered      2017-08-10 21:21:07          Sao Paulo   
115032    delivered      2018-01-08 21:28:27     Rio de Janeiro   
115033    delivered      2018-01-08 21:28:27     Rio de Janeiro   

        payment_sequential payment_type  payment_installments  payment_value  \
12                     1.0  credit_card                   1.0         259.06   
13                     1.0  credit_card                   1.0         259.06   
57                     1.0  credit_card                   2.0         151.04   
58                     1.0  credit_card                   2.0         151.04   
63                     1.0  credit_card                   9.0          91.68   
...                    ...          ...                   ...            ...   
114985                 1.0  credit_card                   9.0         239.50   
115003                 1.0  credit_card                  10.0         748.24   
115004                 1.0  credit_card                  10.0         748.24   
115032                 1.0  credit_card                   4.0         441.16   
115033                 1.0  credit_card                   4.0         441.16   

                              product_id       product_category  
12      08574b074924071f4e201e151b152b4e           garden_tools  
13      08574b074924071f4e201e151b152b4e           garden_tools  
57      f48eb5c2fde13ca63664f0bb05f55346         sports_leisure  
58      f48eb5c2fde13ca63664f0bb05f55346         sports_leisure  
63      b37b72d5a56f887725c2862184b8cab8              telephony  
...                                  ...                    ...  
114985  4f18ca9862f511ecba98258b2194d061      furniture_bedroom  
115003  0df37da38a30a713453b03053d60d3f7         sports_leisure  
115004  0df37da38a30a713453b03053d60d3f7         sports_leisure  
115032  d1c427060a0f73f6b889a5c7c61f2ac4  computers_accessories  
115033  d1c427060a0f73f6b889a5c7c61f2ac4  computers_accessories  

[17347 rows x 11 columns]

In [40]:
# Removing duplicates values
revenue_products = revenue_products.drop_duplicates(keep='first') # Remove duplicate data and keep the first data that appears
revenue_products = revenue_products.reset_index(drop = True) # Reset the column index
revenue_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103172 entries, 0 to 103171
Data columns (total 11 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   order_id                  103172 non-null  object        
 1   customer_id               103172 non-null  object        
 2   order_status              103172 non-null  object        
 3   order_purchase_timestamp  103172 non-null  datetime64[ns]
 4   customer_state            103172 non-null  object        
 5   payment_sequential        103172 non-null  float64       
 6   payment_type              103172 non-null  object        
 7   payment_installments      103172 non-null  float64       
 8   payment_value             103172 non-null  float64       
 9   product_id                103172 non-null  object        
 10  product_category          103172 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 8.7+ MB

In [41]:
def calculate_revenue_per_product_cat(revenue_data, state_list):
    revenue_by_state = {}
    
    for state in state_list:
        state_data = revenue_data[revenue_data['customer_state'] == state]
        revenue_by_state[state] = state_data.groupby('product_category')['payment_value'].sum().reset_index()
        
    return revenue_by_state

def calculate_revenue_proportion(revenue_by_state_dict, state):
    state_summary = revenue_by_state_dict.get(state)

    if state_summary is not None:
        state_summary['revenue_proportion'] = (state_summary['payment_value'] / state_summary['payment_value'].sum()) * 100
    
    return state_summary.rename(columns={'payment_value': 'revenue', 'revenue_proportion': 'revenue_proportion'})

def sort_by_revenue(revenue_summary, ascending=True):
    return revenue_summary.sort_values(by='revenue', ascending=ascending).head(10).reset_index(drop=True)

# Top 5 states with highest revenue
states = ['Sao Paulo', 'Rio de Janeiro', 'Minas Gerais', 'Rio Grande do Sul', 'Parana']

# Calculate revenue per product category for top 5 states
revenue_by_state_dict = calculate_revenue_per_product_cat(revenue_products, states)

In [42]:
# Example usage for sorting by highest revenue
state = 'Sao Paulo'
state_summary = calculate_revenue_proportion(revenue_by_state_dict, state)
highest_SP = sort_by_revenue(state_summary, ascending=False)
highest_SP


product_category    revenue  revenue_proportion
0         bed_bath_table  646481.77           10.616790
1          health_beauty  534209.43            8.773008
2          watches_gifts  466060.17            7.653832
3         sports_leisure  450930.65            7.405369
4  computers_accessories  416581.36            6.841271
5        furniture_decor  374931.61            6.157282
6             housewares  343933.94            5.648225
7                   auto  248547.31            4.081746
8             cool_stuff  235829.18            3.872884
9                   toys  219047.24            3.597284

In [43]:
state = 'Rio de Janeiro'
state_summary = calculate_revenue_proportion(revenue_by_state_dict, state)
highest_RJ = sort_by_revenue(state_summary, ascending=False)
highest_RJ

product_category    revenue  revenue_proportion
0         bed_bath_table  210439.76            9.771637
1          watches_gifts  199439.29            9.260837
2          health_beauty  164879.49            7.656075
3  computers_accessories  149665.28            6.949612
4         sports_leisure  144618.68            6.715277
5        furniture_decor  130653.21            6.066799
6             housewares   96674.72            4.489029
7             cool_stuff   93792.41            4.355191
8           garden_tools   90190.00            4.187916
9                   toys   85233.99            3.957786

In [44]:
state = 'Minas Gerais'
state_summary = calculate_revenue_proportion(revenue_by_state_dict, state)
highest_MG = sort_by_revenue(state_summary, ascending=False)
highest_MG

product_category    revenue  revenue_proportion
0         bed_bath_table  180645.67            9.480303
1          health_beauty  179475.38            9.418886
2          watches_gifts  135518.20            7.112009
3  computers_accessories  133615.01            7.012129
4         sports_leisure  132687.85            6.963472
5        furniture_decor  114502.00            6.009076
6             housewares   98471.07            5.167772
7                   auto   86839.19            4.557329
8             cool_stuff   81562.20            4.280392
9           garden_tools   77931.81            4.089869

In [45]:
state = 'Rio Grande do Sul'
state_summary = calculate_revenue_proportion(revenue_by_state_dict, state)
highest_RS = sort_by_revenue(state_summary, ascending=False)
highest_RS

product_category   revenue  revenue_proportion
0         bed_bath_table  84737.34            9.402476
1        furniture_decor  74042.52            8.215776
2  computers_accessories  66849.60            7.417648
3         sports_leisure  62281.37            6.910756
4          health_beauty  61151.37            6.785371
5          watches_gifts  53104.16            5.892450
6             housewares  51720.12            5.738877
7             cool_stuff  50295.08            5.580754
8           garden_tools  43360.05            4.811242
9                   baby  34073.33            3.780785

In [46]:
state = 'Parana'
state_summary = calculate_revenue_proportion(revenue_by_state_dict, state)
highest_PR = sort_by_revenue(state_summary, ascending=False)
highest_PR

product_category   revenue  revenue_proportion
0        furniture_decor  73966.95            9.022799
1         sports_leisure  68886.50            8.403064
2          watches_gifts  63847.51            7.788387
3         bed_bath_table  63518.33            7.748233
4          health_beauty  62190.89            7.586306
5  computers_accessories  53271.93            6.498334
6             cool_stuff  43456.38            5.300992
7             housewares  40221.13            4.906342
8                   auto  34712.35            4.234358
9                   toys  27932.26            3.407294

## Visualization & Explanatory Analysis

### Pertanyaan 1:
Negara bagian mana saja yang memiliki pendapatan tertinggi?

In [47]:
# Create viz1 dataframe for visualization of revenue by state 
viz1 = revenue_by_state

# Create the visualization of revenue by state
fig = px.bar(data_frame=viz1, x='State', y='Revenue', title='Negara dengan Revenue Tertinggi')
fig.show()

### Pertanyaan 2: 
Bagaimana Trend Pendapatan/Pemasukan di 5 negara dengan revenue tertinggi ?

In [48]:
fig = make_subplots(rows=3, cols=2, subplot_titles=('Revenue Trend Sao Paulo', 'Revenue Trend Rio de Janeiro', 
'Revenue Trend Minas Gerais', 'Revenue Trend Rio Grande do Sul', 'Revenue Trend Parana'))

# Add Plotly Express plots to each subplot
fig.add_trace(px.line(monthly_revenue_SP, x='year_month', y='revenue').data[0], row=1, col=1)
fig.add_trace(px.line(monthly_revenue_RJ, x='year_month', y='revenue').data[0], row=1, col=2)
fig.add_trace(px.line(monthly_revenue_MG, x='year_month', y='revenue').data[0], row=2, col=1)
fig.add_trace(px.line(monthly_revenue_RS, x='year_month', y='revenue').data[0], row=2, col=2)
fig.add_trace(px.line(monthly_revenue_PR, x='year_month', y='revenue').data[0], row=3, col=1)

# Update subplot titles (optional)
fig.update_layout(title_text='Trend Revenue di 5 Negara Bagian',
                  autosize=False,
                  width=1000,
                  height=1200,)

# Show the plot
fig.show()

### Pertanyaan 3: 
Produk apa saja yang menjadi sumber pendapatan terbesar di lima negara bagian tersebut ?

In [49]:
from plotly.subplots import make_subplots
import plotly.express as px

# Daftar summary dan judul subplot
summaries = [highest_SP, highest_RJ, highest_MG, highest_RS, highest_PR]
titles = ['Sumber Pendapatan tertinggi Sao Paulo', 'Sumber Pendapatan tertinggi Rio de Janeiro', 'Sumber Pendapatan tertinggi Minas Gerais',
          'Sumber Pendapatan tertinggi Rio Grande do Sul', 'Sumber Pendapatan tertinggi Parana']

# Membuat subplot
fig = make_subplots(rows=3, cols=2, subplot_titles=titles)

# Menambahkan Plotly Express plots ke setiap subplot
for i, summary in enumerate(summaries):
    row = i // 2 + 1  # Hitung baris subplot
    col = i % 2 + 1   # Hitung kolom subplot
    fig.add_trace(px.bar(summary, x='product_category', y='revenue').data[0], row=row, col=col)

# Memperbarui layout
fig.update_layout(title_text='Sumber Pendapatan terbesar di 5 Negara dengan Revenue Tertinggi',
                  autosize=False,
                  width=1000,
                  height=1200)

# Menampilkan plot
fig.show()

## Conclusion

Negara dengan revenue/pendapatan tertinggi jika diurutkan *descending* adalah Sao Paulo, Rio de Janeiro, Minas Gerais, Rio Grande de Sul, Parana.

Pendapatan di negara bagian Sao Paulo, Rio de Janeiro, Minas Gerais, Rio Grande do Sul, dan Parana menunjukkan tren positif yang cenderung naik secara fluktuatif mulai dari bulan Oktober 2016 hingga Agustus 2018

Sumber pendapatan tertinggi di negara bagian Sao Paulo, Rio de Janeiro, Minas Gerais, Rio Grande do Sul adalah bed_bath_table sedangkan di negara bagian Parana pendapatan tertinggi ada di furniture_decor